added ability to change remote options on the fly

DiliBau opened the pull request:
When trying to validated a field together with an adjacent field, the `parsley:field:validate` is simply not powerful enough to allow you to send additional data dynamically and neither is the `data-parsley-remote-options` attribute.

This applies, for example, when remotely validating international phone numbers which require the user to also select the country in order to apply formatting rules.

```javascript
// this fails and data is never sent
$.listen('parsley:field:validate', function (field) {
  if (field === someRefToMyField) {
    field.options.remoteOptions.data = {someName: $someField.val()};
  }
});
// this works
$.listen('parsley:remote:options', function (field, options) {
  if (field === someRefToMyField) {
    options.data = {someName: $someField.val()};
  }
});
```

marcandre commented:
Thanks for the PR.

An easier approach might be to allow `remoteOptions` to be a function, to be called before an ajax call. There are other requests like #627 to tweak the URL, so the solution needs some reflexion.

DiliBau commented:
@marcandre since the hash that is passed to this callback is the options object that is sent to the `$.ajax()` call, the `url` could easily be changed by doing something like
```javascript
$.listen('parsley:remote:options', function (field, options) {
  if (field === someRefToMyField) {
    options.url = field.$element.data().validationUrl || someFunctionToGenerateTheURL();
  }
});
```

guillaumepotier commented:
I kinda like this approach since it is very lightweight in term or code and allow various tweaks. This seems a bit more like a 'hack' or a 'patch' and as mentioned in #627 proper 'native' implementation should be addressed, meantime this solution could do the job IMHO.

This PR though lacks some documentation and tests :)

Best

DiliBau commented:
@guillaumepotier will add tests and docs over the course of today

marcandre commented:
If we're going to add another event for this, I'd call it `'parsley:field:something'` since it's triggered on a field. Maybe `'parsley:field:ajaxoptions'`?

Still, I'm wondering when someone would ever want to bind more than once to that event. If we never bind more than once, it should be an option instead, no?

guillaumepotier commented:
Let's say you have a pseudo field that check the pseudo existence in database (like Twitter username on signup), using a REST API, successive calls could be

```
GET api.yourservice.com/usernames/foo
GET api.yourservice.com/usernames/foobar
GET api.yourservice.com/usernames/foobarbaz
```

Could be useful to modify on the fly the url on each keypress, no ?

marcandre commented:
Definitely useful!

My question is: when would a given field's ajax options be modified via `'parsley:field:ajaxoptions'` by more than 1 listener? If always 0 or 1 listener, pubsub is overkill and an option `ajaxOptions` would be fine. If sometimes more than 1 listener, then event is much better, as otherwise the inner option `ajaxOptions` has to be aware of the outer one.

guillaumepotier commented:
Ok I see the point now. How could we maybe work that way @DiliBau ?

marcandre commented:
Mmm, I guess a form listener could change the URL in a RESTful way and a local listener could tweak the values according to some kind of formula, so after some more consideration, I'm siding with events after all :smile:

So yeah, the more I think about it, the more awesome that solution sounds. “Only fools never change their mind.”

DiliBau commented:
@guillaumepotier sorry for the delay, i've updated the PR with docs, examples and a test

guillaumepotier commented:
Thx @DiliBau!

Unfortunately, it seems there is a problem with the test suite.

```
browserify test/env/browserify.js > test/env/browserify_bundle.js
Error: Cannot find module 'jquery' from '/home/travis/build/guillaumepotier/Parsley.js/dist'
    at /home/travis/.nvm/v0.10.38/lib/node_modules/browserify/node_modules/resolve/lib/async.js:46:17
    at process (/home/travis/.nvm/v0.10.38/lib/node_modules/browserify/node_modules/resolve/lib/async.js:173:43)
    at ondir (/home/travis/.nvm/v0.10.38/lib/node_modules/browserify/node_modules/resolve/lib/async.js:188:17)
    at load (/home/travis/.nvm/v0.10.38/lib/node_modules/browserify/node_modules/resolve/lib/async.js:69:43)
    at onex (/home/travis/.nvm/v0.10.38/lib/node_modules/browserify/node_modules/resolve/lib/async.js:92:31)
    at /home/travis/.nvm/v0.10.38/lib/node_modules/browserify/node_modules/resolve/lib/async.js:22:47
    at Object.oncomplete (fs.js:108:15)
```

marcandre commented:
May be due to recent change of jQuery path: 4067c4b5ff5b16e6

marcandre commented:
Test looks great!

Couple of things I'd change:
- `parsley:remote:options` is not the right name (as I mentionned before), it should start with `parsley:field:`. I recommend `parsley:field:ajaxoptions`, or maybe even `parsley:field:ajax` since we could provide the possibility to the user to make the ajax call itself and return the promise?
- the PR shouldn't make any changes to `dist/`, that happens only when we make a release
- triviality, but it took me a minute to notice the difference between the two urls in the test, would be best to make them more different than two letters being swapped.

marcandre commented:
Couple more things:
* would be best to remove the custom validator after, to avoid potential interactions with other tests
* in the test, the listener should be added to the `parsleyInstance`, not as a global listener.
* it's not obvious why the validator fails, better `return false`, or why not return `true` and use `done`?

DiliBau commented:
@marcandre @guillaumepotier 
- renamed event to `field:remote.ajaxOptions`  to be consistent with the other events
- reverted changes to `dist`
- updated events to use `on()`, `off()` instead of global `$.listen()`
- updated docs
- fixed the test
- using done to avoid this
```
timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.
```

marcandre commented:
Good work!
I'm confused though: did you remove your `mycustom` validator somehow? I just meant you should call `removeAsynchValidator`. But it looks like that method is missing! So simply use a very specific validator name, we might get around fixing that at some point.

More nitpicking: for reasons I can't fathom, JS events seem to be all lowercase, e.g. `dragend`, `gotpointercapture`, so I wonder if we should follow the (ugly) convention and use `ajaxoptions` instead of `ajaxOptions`.

DiliBau commented:
@guillaumepotier if you want the code, you can find it in my fork, you're free to take it and include it in the main report and/or modify it as you see fit
the example that i included is clear enough to get you started and the tests are passing, other then that i do not have enough time to make @marcandre happy

marcandre commented:
This is now merged in `master`. There's also a simpler way to use the input value in the URL
